common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,313 @@
+
+    /** Failure modes, which will only occur if there is a bug in this implementation */
+    case class ConcurrentOperationUnderRead(actualState: State) extends Exception(s"Cache bug: a read started, but completion raced with a concurrent operation: ${actualState}")
+    case class ConcurrentOperationUnderUpdate(actualState: State) extends Exception("Cache bug: an update started, but completion raced with a concurrent operation: ${actualState}")
+
+    /**
+     * The entries in the cache will be a triple of (transid, State, Future[W])
+     *
+     * We need the transid in order to detect whether we have won the race to add an entry to the cache
+     *
+     */
+    private case class Entry(transid: TransactionId, state: AtomicReference[State], value: Future[W]) {
+        def invalidate = state.set(InvalidateInProgress)
+
+        def writeDone()(implicit logger: Logging): Boolean = {
+            logger.debug(this, "Write finished")
+            trySet(WriteInProgress, Cached)
+        }
+
+        def readDone()(implicit logger: Logging): Boolean = {
+            logger.debug(this, "Read finished")
+            trySet(ReadInProgress, Cached)
+        }
+
+        def trySet(expectedState: State, desiredState: State): Boolean = {
+            state.compareAndSet(expectedState, desiredState)
+        }
+    }
+
+    protected def cacheInvalidate(key: Any)(
+        implicit transid: TransactionId, logger: Logging): Unit = {
+        if (cacheEnabled) {
+            logger.info(this, s"invalidating $key")
+            cache remove key
+        }
+    }
+
+    protected def cacheLookup[Wsuper >: W](
+        key: Any,
+        future: => Future[W],
+        fromCache: Boolean = cacheEnabled)(
+            implicit ec: ExecutionContext, transid: TransactionId, logger: Logging): Future[W] = {
+
+        if (fromCache) {
+            val promise = Promise[W]
+            val desiredEntry = Entry(transid, ReadInProgress, promise.future)
+
+            //
+            // try inserting our desired entry...
+            //
+            cache(key)(desiredEntry) flatMap { actualEntry =>
+                // ... and see what we get back
+
+                actualEntry.state.get match {
+                    case Cached => {
+                        logger.debug(this, s"Cached read $key")
+                        makeNoteOfCacheHit(key)
+                        actualEntry.value
+                    }
+
+                    case ReadInProgress => {
+                        if (actualEntry.transid == transid) {
+                            logger.debug(this, "Read initiated");
+                            makeNoteOfCacheMiss(key)
+                            listenForReadDone(key, actualEntry, future, promise)
+                            actualEntry.value
+
+                        } else {
+                            logger.debug(this, "Coalesced read")
+                            makeNoteOfCacheHit(key)
+                            actualEntry.value
```

Comment:
`actualEntry.value` can be pulled out of the `if ... else ...`, reducing duplication.